torch/csrc/jit/script/compiler.cpp

```diff
@@ -899,13 +932,127 @@ struct to_ir {
     }
   }
 
+  RefinementSet findIsNoneRefinements(
+      Expr lhs,
+      Value* lhs_value,
+      Expr rhs,
+      Value* rhs_value,
+      int tok) {
+    if (rhs.kind() != TK_NONE && lhs.kind() == TK_NONE) {
+      // make 'None is var' into 'var is None'
+      return findIsNoneRefinements(rhs, rhs_value, lhs, lhs_value, tok);
+    }
+    if (rhs.kind() != TK_NONE || lhs.kind() != TK_VAR) {
+      return {};
+    }
+    // statement must be var {is, is not} None
+    auto name = Var(lhs).name().name();
+    // XXX - while it should in theory be possible to specialize
+    // the `x is None` to know x has type NoneType, we have previously not
+    // done this. Unfortunately, doing this will make the type None
+    // propagate further in all loaded models. The handling of
+    // unwrap_optional will fail in these cases since export did
+    // not expect that the input would be none and an unannotated None.
+    // cannot be passed to unwrapoptional To enable this,
+    // we need to (1) implement a real casting operator
+    // annotated(T, X) that stays in the graph and does the cast
+    // and (2) only enable this OPTIONAL_NONE when loading newer
+    // graphs because it is incompatible with older graphs.
+    // Refinement none(name, RefinementKind::OPTIONAL_NONE);
+
+    Refinement present(
+        name, lhs_value->type()->expect<OptionalType>()->getElementType());
```

Comment:
Probably don't want to `expect<OptionalType>` here because it will error for something like:
```
a: int = 1
if a is None:
```

which can happen with a constant attribute that is initialized non-null